### ISSUE TITLE
Add metadata for $PnE (or implement a way to linearize legacy analog data)

### ISSUE THREAD

**Issue by jfx319** [bpteague/cytoflow#178]
https://github.com/bpteague/cytoflow/blob/71e6d9a47ce6b77251553dac9ece89712ae4cf8d/cytoflow/operations/import_op.py#L385

Why not store all the metadata that fcsparser.parse would automatically get somewhere (e.g. `experiment.fcsparse_meta`) in case someone wants to access that info? (as opposed to selectively keeping a few in custom variables). After all, this header info isn't all that big compared to the data values itself; someone may find something useful down the line, and this would be in keeping with the design principle of not losing the original data. 

In my case, I'm working with FCS files that have nonzero values in $PnE (run on machines that stored data values as power 10 transformed according to $PnE and $PnR). In order to convert these back to the linear fluorescence intensity measurements, one would have to use the $PnE exponents and follow equation specified on page 23 of the [FCS3.1 standard](http://isac-net.org/Resources/Standards/FCS3-1.aspx)

However, $PnE is not stored by the import_op, so there's no way to access this info without rerunning fcsparser.parse

I guess the *real* wishlist is to have a built-in way of handling nonzero $PnE (but that's a bigger ask... :P). 

Example flowCore implementation:  
https://github.com/RGLab/flowCore/blob/05e32ea0ec9bb0f3a10e82b048112e23b43d0d5e/R/IO.R#L953

Example python implementation: 
```python
def linearize(data, f1=0, f2=1, PnR=None):
    if (f2 == 0) & (f1 > 0):
        f2 = 1                             #for legacy FCS that have been mislabeled
    return f2*10**(f1*data/PnR)
```

**Comment by bpteague** [comment]
Hi @jfx319 -- thanks for the request.  I may be able to sneak this into the 0.9 release -- can you point me to a file or two with a non-zero $PnE?  And which instrument are you using to generate this data?

**Comment by bpteague** [comment]
(And TBH I'd rather handle $PnE properly.  My goal is to move people *away* from FCS files and tubes after the Import operation.  Which tube's metadata should I store?  All of them?  Where should I store it?  I suppose I could stuff it all into Experiment.metadata... thoughts?)

**Comment by jfx319** [comment]
That example came from [my exploration](https://www.rdocumentation.org/packages/flowCore/versions/1.38.2/topics/spillover) of the default sample data that comes with flowCore: https://github.com/RGLab/flowCore/tree/trunk/inst/extdata

I seems to me that in general, older machines generate nonzero PnE values and newer machines don't (due to analog vs digital hardware transforms). However, public data (e.g. on flowrespository.org) may include older files so would be nice to do the conversion specified in FCS standard, if they're importing an older file. 

Considerations: 
 - I don't know whether this should be something in @eyurtsev 's [fcsparser](https://github.com/eyurtsev/fcsparser) since other pythonic flow utilties, including cytoflow, calls that package; hence, I only felt appropriate asking to keep metadata
 - Experiment.metadata sounds fine (anywhere reasonable, really... keeping in mind that each fcs file may differ slightly in its metadata)
 - I wholeheartedly agree with moving away from FCS files (it does seem outdated since much of the design considerations was for a time when data storage is expensive); other fields are moving towards hdf5 for example); but there's a paucity of python tools for handling fcs (therefore, both cytoflow and flowcytometrytools are appreciated!)
 - In fact, I would move away from cytometry period. The real value in these domain-specific tools that wrap existing datascience tools is that they've implemented a *working* workflow. One would not need a new tool for this (though tools are appreciated), even an up-to-date ipynb that shows how to use existing tools would suffice. Likewise, though docker containers are appreciated, it's really the [recipe file](https://hub.docker.com/r/jfx319/cytometry-notebook/~/dockerfile/) that's more generally useful as they've distilled the trial-and-error hardwork of getting install to work). 
    - The interactive visualization field is developing rapidly; ideally, any gating methods that wrap these tools should be generalizable to other fields (not just flow cytometry)... i.e. in the long run, the ideal situation is to couple fcsparser to move data into a csv.gz or something appropriate and leverage the expanding set of tools in this area: e.g. ipywidgets, altair, bqplot, etc, and either host it online as in binder, or d3.express (now observable), or publish it as a standalone html for non-coding collaborators to interactively play with the charts (e.g. jupyter dashboards)
 - I would love to be a part of creating python or d3 workflows for general data exploration (including the "gating" tool from cytometry field); e.g. [seurat's do.identify example](http://satijalab.org/seurat/faq)

**Comment by bpteague** [comment]
@jfx319 - I have updated `ImportOp` to save each FCS file's metadata in `Experiment.metadata['fcs_metadata']`.  I'm going to leave the bug open so I don't forget to properly handle $PnE, but it's not going to make it into 0.9.

As for your last few comments -- I think you slightly misunderstood me.  When people are analyzing cytometry data, I don't want them to have to think about tubes during their analysis -- just their experimental variables.  Set up the mapping from variables to tubes in `ImportOp` and then forget about it.  Is the FCS standard old and creaky and occasionally busted?  Absolutely.  But as long as we have cytometers, we're going to have FCS files.

Speaking more broadly, I think there *is* still a role for domain-specific tools, because that allows you to take advantage of the structure and regularity of a particular domain's data.  I think this is particularly important for non-programmers, who are going to be less inclined to open a Jupyter notebook, even a pre-populated one.  There's always a tradeoff between power/generality and ease of use.  Cytoflow's intent is to make quantitative analysis of cytometry data accessible to both programmers and non-programmers, and I think in this instance the best way to do that is to sacrifice generality in the name of access and usability.

....and, importantly, then provide an easy path toward more power.  The GUI not doing enough for you? Click the button to export a Jupyter notebook, suck out the underlying `pandas.DataFrame`, and go to town.

**Comment by bpteague** [comment]
(Also, in case it wasn't obviously ... pull requests gladly considered. (-: )

**Comment by jfx319** [comment]
`Experiment.metadata['fcs_metadata']` works great -- thanks! 

if I end up with some free time, might take a stab at generalizing the `def linearize` function above... (I wonder if it's better to implement in cytoflow, or in fcsparser?)

**Comment by bpteague** [comment]
I think cytoflow.operations.import_op would be a good place to do so -- fcsparser is just that, a parser for FCS files.  You could definitely ping @eyurtsev and see if he has strong feelings on the matter, though.

Also -- you didn't have any trouble installing 0.9.1?  Are you using the anaconda package or the pip package?

**Comment by bpteague** [comment]
Hi @jfx319  - I've fixed this in HEAD.  Should make its way into a released version (hopefully 1.0!!) in a couple of weeks.
